### ISSUE TITLE
Excessive stack depth comparing types from conditional types

### ISSUE THREAD

**Issue by dyst5422** [microsoft/TypeScript#50047]
# Bug Report

### 🔎 Search Terms

- 'Excessive stack depth comparing types'
- stack size
- conditional types

### 🕗 Version & Regression Information

- This is a failure to compile

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/dev/bug-workbench/?#code/C4TwDgpgBAwg9gOwCYEtgsQQwDYBVwQA8AglBAB7ATIDOUCArgLYBGEATgHxQC8AUFCikKVWlAAMUAPxQA3lADkALw5wFALkUr2aqAF8o6gULKVqSOgEZpcxYgga7CB-sPHhZsQCYb8hcAB3NU1-IIVXI0EPUQsoAGZfRWAAC3YIBxCUtJcDSJMRczoAFkSFADM4BnZHcsrqiPdTGLoAVlKylAA3DMUO7vDcxoKxADZSmhRyGomphqimwqgAdnGIboRpteoBt3nh2IAOUogUAHNk4BqT88u5-M9YgE5ShBRnGtf3u+jFy0kZPyiGpA74LMSWawAxQQbBbDYhGFwnZ5H7gnxQ0Iw-qZAJYnK7e7NKCWBIYlIodhUbaZZIUqkbUH7KwlDEVKr0mpsynpBmDPYPKxtVkoMockIdUU85FDAXEsYYmZixSKqWM2WWFYKuFKhQ0bWqvmE35HDHXC5ShFnc3bNVEyzPDGfCA6p0c22LLz-WyYhCgYG430gaX8oleSHewLUUAAWns-qjIFjX0NqNiXnREYDMcCwSSWcTOeDRu8pMzCejWXS8cDFdSVfd3hZZZrXOrMdbDbTQub7a6PR9vf6nboXnlPcTMzbE8mRdTI814+jevWU6XSOHUC8JsXZsuOPLu9nYLTDsXTtX543cS9gNplKDNLpQav4dvT6T-fJ94-R6Z8Qzb7foWj5AWEV6loB2Z1p+d5Qdkv6ynETaQYmrYge2dQIUScTdih0Z9DB74EVhixxGOeGTuh06zCmx50HEC4UUiVFriuV7bnhh4sYeV6nnh54sZetF-kUN69HA96chJfobkUr7id+cbitJiZxrJAEKdmYTKaBujCbKRQQZpBbQVJQGmbJyHGfhmE6RhVQkWIRS4dZxF2ahfaObERTkdZlF+TOsmMX5zEhWx+lEkUHHWVxMVWrcEWLEUfHWQJqVvPiKJ0VALRieUIp+uKBXPolYgtPJ+Wiqp7xFVVP4bi0GmVVpubNQWYGlbELRGW1tbwbVcH1p1rRWb1aG9MVNkOQ1LljZ5A0eUOw05b5vX+WtgXLS0wVraFu3hQSc45dFvWxad8Vea0KW9WlN0ZZdUAjHlioPsqkwyctIwVS9P4hD9amfU1P3AW9lDtXph3ZSMPXA6Zf3vSZ8EbiMo0-eNuoI1N9SfbNaPzaDg6ZTKRIjKtP3reTm2Q3+Iw7eTe30wdWU0ydP1nWzF3I9dP23Tz90bksz3aq9urCw9SzfcLv3KlLAPU7KSxA1LIOi+sLXizDytwzLauI0N8tEksqNS+jy7lh2y1LLjJv46r5ueQLZNSxTztU8zCt087DNe0zxOLEsrNS+zQec5b3NS7zEf88tBx5buIvxw9BwVfH0sKKncvu0SBxNanKt5x1BuLAcPV59r6fxXrRMhsXo2p+j9eYRuByzfXttt0tRdiAcq2p+tfdu373c7X3e2j77NfdydqdnTPodd4c12p7dy-RwvdCPHlroi9vD2PBV29p4fmdD08TWHyrF+F1niyPD1F-lw-SPLY8o2H+j79Ny-s3v7bv+dzfMQjxVqH3WqAwek8ng7VAXtGBE9ixPBOofM6yD555GcN0dgABuPguC+BlAYAgAAxugRAUAmAgAAGKEKISQbKjBWAcE4AACkwJoYgAAaKALBNDwGQGgDACAcD4EgCQTgABKXhiBUCkKEXgAgYi5DGDSMAKoCBuE4L0Lgih1DiHMMsFwnRND9FcL8EpJwORxHiKAA)

### 💻 Code

```ts
type ConditionalType<A extends number> =
  A extends 0 ? { 'zero': 'zero' } :
  A extends 1 ? { 'one': 'one' } :
  A extends 2 ? { 'two': 'two' } :
  A extends 3 ? { 'three': 'three' } :
  A extends 4 ? { 'four': 'four' } :
  A extends 5 ? { 'five': 'five' } :
  A extends 6 ? { 'six': 'six' } :
  A extends 7 ? { 'seven': 'seven' } :
  A extends 8 ? { 'eight': 'eight' } :
  A extends 9 ? { 'nine': 'nine' } :
  A extends 10 ? { 'ten': 'ten' } :
  A extends 11 ? { 'eleven': 'eleven' } :
  A extends 12 ? { 'twelve': 'twelve' } :
  A extends 13 ? { 'thirteen': 'thirteen' } :
  A extends 14 ? { 'fourteen': 'fourteen' } :
  A extends 15 ? { 'fifteen': 'fifteen' } :
  A extends 16 ? { 'sixteen': 'sixteen' } :
  A extends 17 ? { 'seventeen': 'seventeen' } :
  A extends 18 ? { 'eighteen': 'eighteen' } :
  A extends 19 ? { 'nineteen': 'nineteen' } :
  A extends 20 ? { 'twenty': 'twenty' } :
  A extends 21 ? { 'twenty-one': 'twenty-one' } :
  A extends 22 ? { 'twenty-two': 'twenty-two' } :
  A extends 23 ? { 'twenty-three': 'twenty-three' } :
  A extends 24 ? { 'twenty-four': 'twenty-four' } :
  A extends 25 ? { 'twenty-five': 'twenty-five' } :
  A extends 26 ? { 'twenty-six': 'twenty-six' } :
  A extends 27 ? { 'twenty-seven': 'twenty-seven' } :
  A extends 28 ? { 'twenty-eight': 'twenty-eight' } :
  A extends 29 ? { 'twenty-nine': 'twenty-nine' } :
  A extends 30 ? { 'thirty': 'thirty' } :
  A extends 31 ? { 'thirty-one': 'thirty-one' } :
  A extends 32 ? { 'thirty-two': 'thirty-two' } :
  A extends 33 ? { 'thirty-three': 'thirty-three' } :
  A extends 34 ? { 'thirty-four': 'thirty-four' } :
  A extends 35 ? { 'thirty-five': 'thirty-five' } :
  A extends 36 ? { 'thirty-six': 'thirty-six' } :
  A extends 37 ? { 'thirty-seven': 'thirty-seven' } :
  A extends 38 ? { 'thirty-eight': 'thirty-eight' } :
  A extends 39 ? { 'thirty-nine': 'thirty-nine' } :
  A extends 40 ? { 'forty': 'forty' } :
  A extends 41 ? { 'forty-one': 'forty-one' } :
  A extends 42 ? { 'forty-two': 'forty-two' } :
  A extends 43 ? { 'forty-three': 'forty-three' } :
  A extends 44 ? { 'forty-four': 'forty-four' } :
  A extends 45 ? { 'forty-five': 'forty-five' } :
  A extends 46 ? { 'forty-six': 'forty-six' } :
  A extends 47 ? { 'forty-seven': 'forty-seven' } :
  A extends 48 ? { 'forty-eight': 'forty-eight' } :
  A extends 49 ? { 'forty-nine': 'forty-nine' } :
  A extends 50 ? { 'fifty': 'fifty' } :
  A extends 51 ? { 'fifty-one': 'fifty-one' } :
  A extends 52 ? { 'fifty-two': 'fifty-two' } :
  A extends 53 ? { 'fifty-three': 'fifty-three' } :
  A extends 54 ? { 'fifty-four': 'fifty-four' } :
  A extends 55 ? { 'fifty-five': 'fifty-five' } :
  A extends 56 ? { 'fifty-six': 'fifty-six' } :
  A extends 57 ? { 'fifty-seven': 'fifty-seven' } :
  A extends 58 ? { 'fifty-eight': 'fifty-eight' } :
  A extends 59 ? { 'fifty-nine': 'fifty-nine' } :
  A extends 60 ? { 'sixty': 'sixty' } :
  A extends 61 ? { 'sixty-one': 'sixty-one' } :
  A extends 62 ? { 'sixty-two': 'sixty-two' } :
  A extends 63 ? { 'sixty-three': 'sixty-three' } :
  A extends 64 ? { 'sixty-four': 'sixty-four' } :
  A extends 65 ? { 'sixty-five': 'sixty-five' } :
  A extends 66 ? { 'sixty-six': 'sixty-six' } :
  A extends 67 ? { 'sixty-seven': 'sixty-seven' } :
  A extends 68 ? { 'sixty-eight': 'sixty-eight' } :
  A extends 69 ? { 'sixty-nine': 'sixty-nine' } :
  A extends 70 ? { 'seventy': 'seventy' } :
  A extends 71 ? { 'seventy-one': 'seventy-one' } :
  A extends 72 ? { 'seventy-two': 'seventy-two' } :
  A extends 73 ? { 'seventy-three': 'seventy-three' } :
  A extends 74 ? { 'seventy-four': 'seventy-four' } :
  A extends 75 ? { 'seventy-five': 'seventy-five' } :
  A extends 76 ? { 'seventy-six': 'seventy-six' } :
  A extends 77 ? { 'seventy-seven': 'seventy-seven' } :
  A extends 78 ? { 'seventy-eight': 'seventy-eight' } :
  A extends 79 ? { 'seventy-nine': 'seventy-nine' } :
  A extends 80 ? { 'eighty': 'eighty' } :
  A extends 81 ? { 'eighty-one': 'eighty-one' } :
  A extends 82 ? { 'eighty-two': 'eighty-two' } :
  A extends 83 ? { 'eighty-three': 'eighty-three' } :
  A extends 84 ? { 'eighty-four': 'eighty-four' } :
  A extends 85 ? { 'eighty-five': 'eighty-five' } :
  A extends 86 ? { 'eighty-six': 'eighty-six' } :
  A extends 87 ? { 'eighty-seven': 'eighty-seven' } :
  A extends 88 ? { 'eighty-eight': 'eighty-eight' } :
  A extends 89 ? { 'eighty-nine': 'eighty-nine' } :
  A extends 90 ? { 'ninety': 'ninety' } :
  A extends 91 ? { 'ninety-one': 'ninety-one' } :
  A extends 92 ? { 'ninety-two': 'ninety-two' } :
  A extends 93 ? { 'ninety-three': 'ninety-three' } :
  A extends 94 ? { 'ninety-four': 'ninety-four' } :
  A extends 95 ? { 'ninety-five': 'ninety-five' } :
  A extends 96 ? { 'ninety-six': 'ninety-six' } :
  A extends 97 ? { 'ninety-seven': 'ninety-seven' } :
  A extends 98 ? { 'ninety-eight': 'ninety-eight' } :
  never;



function myFunc<A extends number>(a: A, b: ConditionalType<A>): ConditionalType<A> {
  return b;
}

myFunc(1, myFunc(1, { 'one': 'one' }))
```

### 🙁 Actual behavior

Compiler failure

```
src/test.ts:213:11 - error TS2321: Excessive stack depth comparing types 'ConditionalType<?>' and 'ConditionalType<?>'.

213 myFunc(1, myFunc(1, { 'one': 'one' }))
              ~~~~~~~~~~~~~~~~~~~~~~~~~~~


Found 1 error in src/test.ts:213
```

### 🙂 Expected behavior

I would expect this to correctly check. In the absence of that, alternative patterns for typing one input based on another that doesn't hit this limit.

**Comment by dyst5422** [comment]
Closing because we found a solution for our use case.

In our case, we went ahead and created a map to the associated types.

So for the above it would look something like

```
type TypeMap = {
  0: {'zero': 'zero'},
  1: {'one': 'one'},
}

function myFunc<A extends keyof TypeMap>(a: A, b: TypeMap[A]): TypeMap[A] {
  return b;
}


myFunc(1, myFunc(1, {'one': 'one'}));
```

TLDR: same programming efficiency stuff you learned in programming 101: use maps, they're great
